### ISSUE TITLE
Passing additional named parameters to a callback/interface is painful

### ISSUE THREAD

**Issue by nex3** [dart-lang/language#1673]
Although I generally find Dart much better suited than JS to publishing re-usable library packages, there's one pattern that JS makes *much* easier:

* I publish a package with an API that users must match. This could be a callback they pass into my code, it could be an interface they implement and pass into my code, whatever.
* Later on, I want to provide extra information to this API without breaking existing users of my library.

In JS, I would simply pass in an extra named parameter (in the form of an extra field on an object). But in Dart, there's no idiomatic way to handle this. There are a few options that aren't great:

* You can simply add the parameter and make it a breaking change, but breaking changes are holistically awful for everyone involved.
* You can add some sort of flag interface that adds the extra parameter, so you can write `if (widget is WidgetWithOption) widget.run(option: value)`, but this is very verbose on both ends and doesn't scale well to multiple additional options.
* You can pass options fully out-of-band using Zone values, but this is highly non-idiomatic and would be very strange from a user's perspective.

**Comment by eernstg** [comment]
It would actually be possible to handle the breaking change of adding a new named parameter in a way that degrades gracefully:

If an instance method declaration `D` named `m` is changed such that a new named parameter `p` is added, every method declaration which is now no more a correct override will be in a library which needs to be recompiled (today, and for any type safe modular compilation technique that I can think of). So we can rely on the fact that the breakage only occurs in code that will be compiled.

A compiler could then emit a diagnostic message for each of those now-incorrect overriding declarations, add the new method parameter, and proceed as if that parameter had been present in the source code already.

There are some details, of course. For a given declaration `D1` that overrides `D` and needs to have `p` added, we could use the type of `p` in `D` as the type of `p` in `D1`. However, we may have multiple declarations of `p` with different (unrelated) types in different superinterfaces (we could have added `p` in more than one location), so it's a more robust approach to always use the parameter type `void` (that parameter will not be used by the body of the method!) and making the parameter non-required.

So in a lot of situations it would just work to let the compiler fix the program. However, there is one possibility for a real conflict: We could have an existing declaration of `p` in an overriding declaration `D1`. This is a conflict that we _must_ report as an error, because this existing `p` could have a completely different purpose than the one that we just added to `D`.

In summary, this means that we could make the breaking change of adding a new named parameter `p` quite "soft": If there are no named parameters in overriding declarations with the same name, then the compilation will emit some diagnostic messages, but compilation can be allowed to succeed and it will just work. It will then be possible to migrate the code (top-down: overridden declarations before overriding ones) at the pace that fits the community.

**Comment by leafpetersen** [comment]
I wonder if another path to help with this is in the tooling.  If we made it really easy for package authors to bundle up an API "quick fix" in their package, then as a package author, you could make the breaking change trivially fixable in most cases (when pub upgrade is run, you get some suggested auto-fixes to apply).  This doesn't help with installed software though - you still get the breakage.  Though perhaps you could just have an option to auto-apply fixes?

@bwilkerson @pq is this something that your API fixup tooling could support?

Another vague possibility (though I have some doubts about it) would be to provide a way to attach a macro to a method which rewrites overrides/implementations (e.g. to add the missing named parameter).  cc @munificent @jakemac53

**Comment by jakemac53** [comment]
If I understand correctly, you could also somewhat mimic the JS approach by passing a single configuration object as a parameter.

You can then add fields to that configuration object without breaking users (assuming they don't implement it). They can both create the object using the old code (assuming you don't add new _required_ params to the constructor), and they can also accept new versions of it in their own apis (and will just ignore any unknown options).

**Comment by nex3** [comment]
That's true, but that's also highly non-idiomatic, and not a pattern you can safely move to from an existing API that already takes named parameters.

**Comment by tatumizer** [comment]
Let's consider a concrete example.
Right now, method `map` of Iterable has a parameter: callback function with a single argument (element). Suppose we want to add another parameter (index) to the callback. We make it a *named* parameter.
My question is: why does this extra *named* parameter break existing programs? Whoever doesn't need this parameter, just ignores it. Is there any logical reason to make it an error?
(I understand the reason for not allowing adding positional parameters, but here I'm interested in named parameters only)

**Comment by jakemac53** [comment]
> My question is: why does this extra _named_ parameter break existing programs? Whoever doesn't need this parameter, just ignores it. Is there any logical reason to make it an error?

The problem is the thing that accepts that callback (Iterable.map) in this case expects to be able to call that function and pass that parameter.

Eriks suggestion above would somewhat resolve this though, you just add essentially a synthetic named parameter to the callback when you pass it in, which is ignored by the actual callback. That does on the surface seem potentially reasonable to me, but it does get weird when you consider dynamic calls. I would assume it would be a static transformation only (you could easily write this today as a kernel transform). That means the same callback can be passed if the function type of the parameter it is being passed for is known statically, but not if it isn't, which is weird. Doing it for dynamic would imply changing the shape of the function on the fly and I don't know if that is something that we could reasonably do on all implementations?

**Comment by tatumizer** [comment]
> The problem is the thing that accepts that callback (Iterable.map) in this case expects to be able to call that function and pass that parameter.

Sure it does expect this, and there's no logical reason for not allowing it to do so. The reason is (probably) a technical one: parameter positions on stack can get confused. I'm pretty sure the problem can be solved assuming the entire program gets recompiled. This shouldn't be a breaking change. The issue is quite consequential IMO.

**Comment by nex3** [comment]
I would expect dynamic calls to be *easier* under Erik's proposal: you'd just define `foo(name: value)` to discard the `name` parameter if `foo` doesn't have one, or cast it to the parameter type if it does.

**Comment by lrhn** [comment]
Discarding the extra argument sounds like JavaScript's approach. It's not a new request (I vaguely remember Angular people asking for being able to pass a one-parameter function to something expecting a 2+ parameter function, and then have a way to invoke it with 2+ parameters safely). 
Simply discarding extra parameters on a dynamic invocation is not something I'd want to do by default. If nothing else, I'd require you to cast the function to a special type, say `Function*`, which has that behavior (otherwise equivalent to `Function`).

For invocation, we could instead add an extra flag to `Function.apply`, say `ignoreExtraArguments: true`, and then it invokes the function with as many arguments as possible from the ones supplied. (Probably going to be at least as expensive as a normal dynamic invocation).

Type-safety is obviously a problem, if you want to accept otherwise incompatible functions (one requiring an extra argument, one not accepting it). It's hard to specify such a parameter type. Imagine (complete strawman) that you could annotate a function type parameter as `hypthetical`, and then it's OK for a function of that type to not have that parameter, but if it has it, it should satisfy the specified type.
```dart
Iterable<S> map<S>(S Function(E element, hypothetical int index) convert) sync* {
   var index = 0;
   for (var element in this) {
     yield Function.apply(convert, [element, index], ignoreExtraArguments: true); 
     index++;
  }
}
```
I'm not sure how subtyping would work out for that. I guess it's really a union type of the possible combinations of hypothetical parameters. (So, if we had union types, it would be `S Function(E)|S Function(E,int)`, and then you'd have to figure out which type it was before calling, which doesn't really scale to multiple hypothetical named parameters, or use a special `Function.apply`).

That still wouldn't allow you to make the change and have subclasses automatically follow through, not unless `hypthetical` parameters are implicitly inherited into the subclass method signature. Even then, if the superclass was the `map` above,
and the existing superclass was:
```dart
  Iterable<S> map<S>(S Function(E) convert) sync* {
    for (var element in this) yield convert(element);
  }
```
then there is no way it can accept an `S Function(E, int)` conversion function, because it has no second argument to call it with.
So, we have to go back and declare with a default value:
```dart

**Comment by eernstg** [comment]
About function literals and dynamic invocations: I proposed a purely static approach, where the semantics of Dart programs would be unchanged, but compilers would perform an autofix operation in the case where a function declaration (be it a function literal or an instance member declaration) has a compile-time error because of a missing named parameter.

As I mentioned, it doesn't handle all cases, but the point is that the migration to the new version of the code degrades gracefully. In some situations it may just continue to work after the breaking change (relying on autofixes), and in other cases there will be some errors that must be handled manually. This allows developers to migrate in a more incremental fashion.

```dart
class A {
  void foo({required int p}) {} // Edit: Added new named parameter `p`.
}

class B extends A {
  void foo() {} // Error here, because of the edit: Handled by autofix.
  // The autofix makes it `void foo({void p}) {}`.
}

class B2 extends A { // Maybe we also have a class like this.
  // If we have a pre-existing `p` then it is likely to have a different purpose.
  void foo({String p = 'We had this parameter for years!'}) {} // Error! Can't autofix.
}

void bar(void Function({int? p}) callback) { // Edit: Added `p`.
  callback(p: 5); // Edit: Using the newly added parameter.
}

void bar2(void Function() callback) { // No edits in signature.
  // Edit the body to use the added parameter if supported.
  if (callback is void Function({int? p}) {
    callback(p: 5);
  } else {
    callback();
  }
}

void main() {
  bar(() {}); // Error here, handled by autofix, `({void p}) {}`.

  var g = () {};
  bar(g); // Error, can't autofix!
  bar(g as Function); // Throws at run time, `g` has a wrong type.
  (g as Function)(p: 5); // Throws at run time, `g` doesn't accept `p`.
}
```

We could of course provide quick fixes in IDEs to perform exactly the same transformations, but there might be a separate value in the "degrade gracefully" property: First, a compile-time autofix can be applied to all libraries in need of fixes, including code that the developer does not maintain. Second, those autofixes are clearly identified as a temporary solution, and it seems likely that those named parameters would be used in some cases, so it makes sense that actual manual edits to the code would be different.

The question is whether it's worthwhile to have such a concept as a compile-time autofix (accompanied by diagnostic messages), just for the purpose of allowing software evolution to proceed more smoothly during migrations in response to this kind of breaking change?

**Comment by tatumizer** [comment]
It's not difficult to construct an example to demonstrate that adding an extra parameter may be a very breaking change.
Consider this function:
```dart
void foo(int Function() giveMeRandomNumber) {...};
```
In the next release we decided to add a parameter maxValue, so that the random number now has to be in a range 0...maxValue:
```dart
void foo(int Function(int maxValue) giveMeRandomNumber) {...};
```
Clearly, all the earlier implementations of `giveMeRandomNumber` become incorrect; the added"maxValue"  parameter changes the meaning of the function. It's difficult (in general- impossible) to fix the code using any autofix. 

On the other hand, adding the "index" parameter to the `Iterable.map` callback is clearly a non-breaking change. 

There's no way for a compiler to figure out which case is which. We need an annotation or a keyword to tell the compiler whether the parameter is "ignorable" for a callee or not. 
Notice that the situation is "symmetric" to the concept of optional parameters.
When the parameter is optional, this by definition means that the *caller* of the function is not required to explicitly pass it while calling the function.
What we are looking for is a similar concept of *callee* optionality, namely: "the callee may ignore this parameter". 
Example:
```dart
Iterable<T> map<T>(T f(E e, ignorable int index));
```
There's nothing radical about the idea - the concept of "callee-optionals" is "symmetric" to the existing concept of "caller-optionals".

**Comment by munificent** [comment]
I agree this is an annoying part of API evolution. I've run into this too.

I'm not personally enthused by the idea of discarding unused function parameters. It makes some changes non-breaking, but can turn other (rarer) changes into breaking ones:

Let's say version one of your API looks like:

```dart
// version 1
doThing(void Function(int i) callback) { callback(0); }
```

I use the API like so:

```dart
takeOneOrTwo(int i, [int? j]) {
  print(i + (j ?? 0));
}

main() {
  doThing(takeOneOrTwo);
}
```

This is allowed since a function with an extra optional parameter is a subtype of a function with fewer parameters. In version two, you change the API to:

```dart
// version 2
doThing(void Function(int i, [String? s]) callback) { callback(0, ""); }
```

Now my code is broken because the optional parameter you added to `doThing()` is a different type from the (ignored) one in the function I was using.

This would definitely be a less common kind of breakage (and using named parameters would make it even rarer), but it's still a thing.

Also, loosening function subtyping to allow dropping unused arguments may mask so many bugs that it cancels out the positive value of the change. Consider cases where you aren't trying to evolve an *external* API in a *non*-breaking way, but simply refactoring your own code. In that case, you *do* want the static error on the extra parameter so that you can find all of those locations in your code and fix them.

There are a couple of (non-ideal) options to work around this:

1. Sometimes you can just add a second "overloaded" method for the new function type. Instead of changing `doThing()` in the above example, add a new `doThingWithTwo()` that takes the updated function type.

2. It's not super common, but I do sometimes use "configuration parameters" like @jakemac53 suggests specifically because they are easier to evolve.

3. You can loosen the API to just `Function` and then test internally to see what kind of function you were given and handle the overloading dynamically.

API evolution is hard.
